File: app/src/main/java/ai/elimu/analytics/task/ExportEventsToCsvWorker.java
package ai.elimu.analytics.task;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.work.Worker;
import androidx.work.WorkerParameters;

import org.apache.commons.csv.CSVFormat;
import org.apache.commons.csv.CSVPrinter;
import org.apache.commons.io.FileUtils;

import java.io.File;
import java.io.IOException;
import java.io.StringWriter;
import java.text.SimpleDateFormat;
import java.util.List;

import ai.elimu.analytics.dao.LetterSoundLearningEventDao;
import ai.elimu.analytics.dao.StoryBookLearningEventDao;
import ai.elimu.analytics.dao.WordAssessmentEventDao;
import ai.elimu.analytics.dao.WordLearningEventDao;
import ai.elimu.analytics.db.RoomDb;
import ai.elimu.analytics.entity.LetterSoundLearningEvent;
import ai.elimu.analytics.entity.StoryBookLearningEvent;
import ai.elimu.analytics.entity.WordAssessmentEvent;
import ai.elimu.analytics.entity.WordLearningEvent;
import ai.elimu.analytics.util.VersionHelper;
import timber.log.Timber;

/**
 * Exports events from the database into CSV files, that will later be uploaded to the server by
 * the {@link UploadEventsWorker}.
 */
public class ExportEventsToCsvWorker extends Worker {

    public ExportEventsToCsvWorker(@NonNull Context context, @NonNull WorkerParameters workerParams) {
        super(context, workerParams);
    }

    @NonNull
    @Override
    public Result doWork() {
        Timber.i("doWork");

        exportLetterSoundLearningEventsToCsv();
        exportWordLearningEventsToCsv();
        exportWordAssessmentEventsToCsv();
        exportStoryBookLearningEventsToCsv();

        return Result.success();
    }

    private void exportLetterSoundLearningEventsToCsv() {
        Timber.i("exportLetterSoundLearningEventsToCsv");

        // Extract LetterSoundLearningEvents from the database that have not yet been exported to CSV.
        RoomDb roomDb = RoomDb.getDatabase(getApplicationContext());
        LetterSoundLearningEventDao letterSoundLearningEventDao = roomDb.letterSoundLearningEventDao();
        List<LetterSoundLearningEvent> letterSoundLearningEvents = letterSoundLearningEventDao.loadAllOrderedByTime();
        Timber.i("letterSoundLearningEvents.size(): " + letterSoundLearningEvents.size());

        CSVFormat csvFormat = CSVFormat.DEFAULT
                .withHeader(
                        "id",
                        "time",
                        "android_id",
                        "package_name",
                        "letter_sound_id",
                        "letter_sound_letter_texts",
                        "letter_sound_sound_values_ipa"
                );
        StringWriter stringWriter = new StringWriter();
        try {
            CSVPrinter csvPrinter = new CSVPrinter(stringWriter, csvFormat);


            // Generate one CSV file per day of events
            String dateOfPreviousEvent = null;
            for (LetterSoundLearningEvent letterSoundLearningEvent : letterSoundLearningEvents) {
                // Export event to CSV file. Example format:
                //   files/version-code-3001017/letter-sound-learning-events/7161a85a0e4751cd_3001017_letter-sound-learning-events_2023-10-25.csv
                Integer versionCode = VersionHelper.getAppVersionCode(getApplicationContext());
                SimpleDateFormat simpleDateFormat = new SimpleDateFormat("yyyy-MM-dd");
                String date = simpleDateFormat.format(letterSoundLearningEvent.time.getTime());
                if (!date.equals(dateOfPreviousEvent)) {
                    // Reset file content
                    stringWriter = new StringWriter();
                    csvPrinter = new CSVPrinter(stringWriter, csvFormat);
                }
                dateOfPreviousEvent = date;
                String csvFilename = letterSoundLearningEvent.androidId + "_" + versionCode + "_letter-sound-learning-events_" + date + ".csv";
                Timber.i("csvFilename: " + csvFilename);

                csvPrinter.printRecord(
                        letterSoundLearningEvent.getId(),
                        letterSoundLearningEvent.time.getTimeInMillis(),
                        letterSoundLearningEvent.androidId,
                        letterSoundLearningEvent.packageName,
                        letterSoundLearningEvent.getId(),
                        null,
                        null
                );
                csvPrinter.flush();

                String csvFileContent = stringWriter.toString();

                // Write the content to the CSV file
                File filesDir = getApplicationContext().getFilesDir();
                File versionCodeDir = new File(filesDir, "version-code-" + versionCode);
                File letterSoundLearningEventsDir = new File(versionCodeDir, "letter-sound-learning-events");
                File csvFile = new File(letterSoundLearningEventsDir, csvFilename);
                FileUtils.writeStringToFile(csvFile, csvFileContent, "UTF-8");
            }
        } catch (IOException e) {
            Timber.e(e);
        }
    }

    private void exportWordLearningEventsToCsv() {
        Timber.i("exportWordLearningEventsToCsv");

        // Extract WordLearningEvents from the database that have not yet been exported to CSV.
        RoomDb roomDb = RoomDb.getDatabase(getApplicationContext());
        WordLearningEventDao wordLearningEventDao = roomDb.wordLearningEventDao();
        List<WordLearningEvent> wordLearningEvents = wordLearningEventDao.loadAllOrderedByTimeDesc();
        Timber.i("wordLearningEvents.size(): " + wordLearningEvents.size());

        CSVFormat csvFormat = CSVFormat.DEFAULT
                .withHeader(
                        "id",
                        "time",
                        "android_id",
                        "package_name",
                        "word_id",
                        "word_text",
                        "learning_event_type"
                );
        StringWriter stringWriter = new StringWriter();
        try {
            CSVPrinter csvPrinter = new CSVPrinter(stringWriter, csvFormat);

            // Generate one CSV file per day of events
            String dateOfPreviousEvent = null;
            for (WordLearningEvent wordLearningEvent : wordLearningEvents) {
                // Export event to CSV file. Example format:
                //   files/version-code-3001012/word-learning-events/7161a85a0e4751cd_3001012_word-learning-events_2020-03-21.csv
                Integer versionCode = VersionHelper.getAppVersionCode(getApplicationContext());
                SimpleDateFormat simpleDateFormat = new SimpleDateFormat("yyyy-MM-dd");
                String date = simpleDateFormat.format(wordLearningEvent.time.getTime());
                if (!date.equals(dateOfPreviousEvent)) {
                    // Reset file content
                    stringWriter = new StringWriter();
                    csvPrinter = new CSVPrinter(stringWriter, csvFormat);
                }
                dateOfPreviousEvent = date;
                String csvFilename = wordLearningEvent.androidId + "_" + versionCode + "_word-learning-events_" + date + ".csv";
                Timber.i("csvFilename: " + csvFilename);

                csvPrinter.printRecord(
                        wordLearningEvent.getId(),
                        wordLearningEvent.time.getTimeInMillis(),
                        wordLearningEvent.androidId,
                        wordLearningEvent.packageName,
                        wordLearningEvent.getWordId(),
                        wordLearningEvent.getWordText(),
                        wordLearningEvent.getLearningEventType()
                );
                csvPrinter.flush();

                String csvFileContent = stringWriter.toString();

                // Write the content to the CSV file
                File filesDir = getApplicationContext().getFilesDir();
                File versionCodeDir = new File(filesDir, "version-code-" + versionCode);
                File wordLearningEventsDir = new File(versionCodeDir, "word-learning-events");
                File csvFile = new File(wordLearningEventsDir, csvFilename);
                FileUtils.writeStringToFile(csvFile, csvFileContent, "UTF-8");
            }
        } catch (IOException e) {
            Timber.e(e);
        }
    }

    private void exportWordAssessmentEventsToCsv() {
        Timber.i("exportWordAssessmentEventsToCsv");

        // Extract WordAssessmentEvents from the database that have not yet been exported to CSV.
        RoomDb roomDb = RoomDb.getDatabase(getApplicationContext());
        WordAssessmentEventDao wordAssessmentEventDao = roomDb.wordAssessmentEventDao();
        List<WordAssessmentEvent> wordAssessmentEvents = wordAssessmentEventDao.loadAllOrderedByTimeAsc();
        Timber.i("wordAssessmentEvents.size(): " + wordAssessmentEvents.size());

        CSVFormat csvFormat = CSVFormat.DEFAULT
                .withHeader(
                        "id",
                        "time",
                        "android_id",
                        "package_name",
                        "word_id",
                        "word_text",
                        "mastery_score",
                        "time_spent_ms"
                );
        StringWriter stringWriter = new StringWriter();
        try {
            CSVPrinter csvPrinter = new CSVPrinter(stringWriter, csvFormat);

            // Generate one CSV file per day of events
            String dateOfPreviousEvent = null;
            for (WordAssessmentEvent wordAssessmentEvent : wordAssessmentEvents) {
                // Export event to CSV file. Example format:
                //   files/version-code-3001012/word-assessment-events/7161a85a0e4751cd_3001012_word-assessment-events_2020-03-21.csv
                Integer versionCode = VersionHelper.getAppVersionCode(getApplicationContext());
                SimpleDateFormat simpleDateFormat = new SimpleDateFormat("yyyy-MM-dd");
                String date = simpleDateFormat.format(wordAssessmentEvent.getTime().getTime());
                if (!date.equals(dateOfPreviousEvent)) {
                    // Reset file content
                    stringWriter = new StringWriter();
                    csvPrinter = new CSVPrinter(stringWriter, csvFormat);
                }
                dateOfPreviousEvent = date;
                String csvFilename = wordAssessmentEvent.getAndroidId() + "_" + versionCode + "_word-assessment-events_" + date + ".csv";
                Timber.i("csvFilename: " + csvFilename);

                csvPrinter.printRecord(
                        wordAssessmentEvent.getId(),
                        wordAssessmentEvent.getTime().getTimeInMillis(),
                        wordAssessmentEvent.getAndroidId(),
                        wordAssessmentEvent.getPackageName(),
                        wordAssessmentEvent.getWordId(),
                        wordAssessmentEvent.getWordText(),
                        wordAssessmentEvent.getMasteryScore(),
                        wordAssessmentEvent.getTimeSpentMs()
                );
                csvPrinter.flush();

                String csvFileContent = stringWriter.toString();

                // Write the content to the CSV file
                File filesDir = getApplicationContext().getFilesDir();
                File versionCodeDir = new File(filesDir, "version-code-" + versionCode);
                File wordAssessmentEventsDir = new File(versionCodeDir, "word-assessment-events");
                File csvFile = new File(wordAssessmentEventsDir, csvFilename);
                FileUtils.writeStringToFile(csvFile, csvFileContent, "UTF-8");
            }
        } catch (IOException e) {
            Timber.e(e);
        }
    }
    
    private void exportStoryBookLearningEventsToCsv() {
        Timber.i("exportStoryBookLearningEventsToCsv");
        
        // Extract StoryBookLearningEvents from the database that have not yet been exported to CSV.
        RoomDb roomDb = RoomDb.getDatabase(getApplicationContext());
        StoryBookLearningEventDao storyBookLearningEventDao = roomDb.storyBookLearningEventDao();
        List<StoryBookLearningEvent> storyBookLearningEvents = storyBookLearningEventDao.loadAll();
        Timber.i("storyBookLearningEvents.size(): " + storyBookLearningEvents.size());

        CSVFormat csvFormat = CSVFormat.DEFAULT
                .withHeader(
                        "id",
                        "time",
                        "android_id",
                        "package_name",
                        "storybook_id",
                        "learning_event_type"
                );
        StringWriter stringWriter = new StringWriter();
        try {
            CSVPrinter csvPrinter = new CSVPrinter(stringWriter, csvFormat);

            // Generate one CSV file per day of events
            String dateOfPreviousEvent = null;
            for (StoryBookLearningEvent storyBookLearningEvent : storyBookLearningEvents) {
                // Export event to CSV file. Example format:
                //   files/version-code-3001012/storybook-learning-events/7161a85a0e4751cd_3001012_storybook-learning-events_2020-03-21.csv
                Integer versionCode = VersionHelper.getAppVersionCode(getApplicationContext());
                SimpleDateFormat simpleDateFormat = new SimpleDateFormat("yyyy-MM-dd");
                String date = simpleDateFormat.format(storyBookLearningEvent.time.getTime());
                if (!date.equals(dateOfPreviousEvent)) {
                    // Reset file content
                    stringWriter = new StringWriter();
                    csvPrinter = new CSVPrinter(stringWriter, csvFormat);
                }
                dateOfPreviousEvent = date;
                String csvFilename = storyBookLearningEvent.androidId + "_" + versionCode + "_storybook-learning-events_" + date + ".csv";
                Timber.i("csvFilename: " + csvFilename);

                csvPrinter.printRecord(
                        storyBookLearningEvent.getId(),
                        storyBookLearningEvent.time.getTimeInMillis(),
                        storyBookLearningEvent.androidId,
                        storyBookLearningEvent.packageName,
                        storyBookLearningEvent.getStoryBookId(),
                        storyBookLearningEvent.getLearningEventType()
                );
                csvPrinter.flush();

                String csvFileContent = stringWriter.toString();

                // Write the content to the CSV file
                File filesDir = getApplicationContext().getFilesDir();
                File versionCodeDir = new File(filesDir, "version-code-" + versionCode);
                File storyBookLearningEventsDir = new File(versionCodeDir, "storybook-learning-events");
                File csvFile = new File(storyBookLearningEventsDir, csvFilename);
                FileUtils.writeStringToFile(csvFile, csvFileContent, "UTF-8");
            }
        } catch (IOException e) {
            Timber.e(e);
        }
    }
}

File: app/src/main/java/ai/elimu/analytics/task/ExportEventsToCsvWorker.kt
package ai.elimu.analytics.task

import ai.elimu.analytics.BuildConfig
import ai.elimu.analytics.db.RoomDb
import ai.elimu.analytics.util.SharedPreferencesHelper
import android.content.Context
import androidx.work.Worker
import androidx.work.WorkerParameters
import org.apache.commons.csv.CSVFormat
import org.apache.commons.csv.CSVPrinter
import org.apache.commons.io.FileUtils
import timber.log.Timber
import java.io.File
import java.io.StringWriter
import java.text.SimpleDateFormat
import java.util.Locale

/**
 * Exports events from the database into CSV files, that will later be uploaded to the server by
 * the [UploadEventsWorker].
 *
 * Folder structure:
 * ```
 * ├── lang-ENG
 * │    ├── letter-sound-learning-events
 * │    │    ├── e387e38700000001_4001000_letter-sound-learning-events_2025-07-27.csv
 * │    │    ├── e387e38700000001_4001000_letter-sound-learning-events_2025-07-27.csv
 * │    │    └── e387e38700000001_4001000_letter-sound-learning-events_2025-07-27.csv
 * │    ├── word-learning-events
 * │    │    ├── e387e38700000002_4000018_word-learning-events_2025-07-27.csv
 * │    │    ├── e387e38700000002_4000018_word-learning-events_2025-07-27.csv
 * │    │    ├── e387e38700000002_4000018_word-learning-events_2025-07-27.csv
 * ```
 */
class ExportEventsToCsvWorker(context: Context, workerParams: WorkerParameters) :
    Worker(context, workerParams) {

    private val eventDateFormat by lazy { SimpleDateFormat("yyyy-MM-dd", Locale.US) }

    override fun doWork(): Result {
        Timber.i("doWork")

        exportLetterSoundLearningEvents()
        exportLetterSoundAssessmentEvents()

        exportWordLearningEvents()
        exportWordAssessmentEvents()

        exportNumberLearningEvents();
        exportNumberAssessmentEvents()

        exportStoryBookLearningEvents()

        exportVideoLearningEvents()

        return Result.success()
    }

    private fun exportLetterSoundLearningEvents() {
        Timber.i("exportLetterSoundLearningEvents")

        // Read all the events from the database
        val roomDb = RoomDb.getDatabase(applicationContext)
        val letterSoundLearningEventDao = roomDb.letterSoundLearningEventDao()
        val events = letterSoundLearningEventDao.loadAllOrderedByTimestampAsc()
        Timber.i("events.size: ${events.size}")

        // Generate one CSV file per day of events, e.g:
        //   lang-THA/letter-sound-learning-events/5b7c682a12ecbe2e_4001000_letter-sound-learning-events_2025-07-27.csv
        //   lang-THA/letter-sound-learning-events/5b7c682a12ecbe2e_4001000_letter-sound-learning-events_2025-07-27.csv
        var stringWriter: StringWriter? = null
        var csvPrinter: CSVPrinter? = null
        var dateOfPreviousEvent: String? = null
        for (event in events) {
            // Get the event's date in ISO format, e.g. "2025-07-27"
            val date: String = eventDateFormat.format(event.timestamp.time)

            // Prepare the CSV file path
            val languageDir = File(applicationContext.filesDir, "lang-${SharedPreferencesHelper.getLanguage(applicationContext)}")
            val eventsDir = File(languageDir, "letter-sound-learning-events")
            val csvFile = File(eventsDir, "${event.androidId}_${BuildConfig.VERSION_CODE}_letter-sound-learning-events_${date}.csv")

            if (date != dateOfPreviousEvent) {
                // Reset file content, and prepare the headers for a new CSV file
                Timber.i("csvFile: ${csvFile}")
                stringWriter = StringWriter()
                csvPrinter = CSVPrinter(stringWriter, CSVFormat.DEFAULT.builder().setHeader(
                    "id",
                    "timestamp",
                    "package_name",
                    "additional_data",
                    "research_experiment",
                    "experiment_group",
                    "letter_sound_letters",
                    "letter_sound_sounds",
                    "letter_sound_id"
                ).get())
            }
            csvPrinter?.printRecord(
                event.id,
                event.timestamp.timeInMillis / 1_000,
                event.packageName,
                event.additionalData,
                event.researchExperiment?.ordinal,
                event.experimentGroup?.ordinal,
                event.letterSoundLetters,
                event.letterSoundSounds,
                event.letterSoundId
            )
            csvPrinter?.flush()

            // Write the content to the CSV file
            val csvFileContent = stringWriter.toString()
            FileUtils.writeStringToFile(csvFile, csvFileContent, "UTF-8")

            dateOfPreviousEvent = date
        }

        Timber.i("exportLetterSoundLearningEvents complete!")
    }

    private fun exportLetterSoundAssessmentEvents() {
        Timber.i("exportLetterSoundAssessmentEvents")

        // Read all the events from the database
        val roomDb = RoomDb.getDatabase(applicationContext)
        val letterSoundAssessmentEventDao = roomDb.letterSoundAssessmentEventDao()
        val events = letterSoundAssessmentEventDao.loadAllOrderedByTimestampAsc()
        Timber.i("events.size: ${events.size}")

        // Generate one CSV file per day of events, e.g:
        //   lang-THA/letter-sound-assessment-events/5b7c682a12ecbe2e_4001000_letter-sound-assessment-events_2025-07-27.csv
        //   lang-THA/letter-sound-assessment-events/5b7c682a12ecbe2e_4001000_letter-sound-assessment-events_2025-07-27.csv
        var stringWriter: StringWriter? = null
        var csvPrinter: CSVPrinter? = null
        var dateOfPreviousEvent: String? = null
        for (event in events) {
            // Get the event's date in ISO format, e.g. "2025-07-27"
            val date: String = eventDateFormat.format(event.timestamp.time)

            // Prepare the CSV file path
            val languageDir = File(applicationContext.filesDir, "lang-${SharedPreferencesHelper.getLanguage(applicationContext)}")
            val eventsDir = File(languageDir, "letter-sound-assessment-events")
            val csvFile = File(eventsDir, "${event.androidId}_${BuildConfig.VERSION_CODE}_letter-sound-assessment-events_${date}.csv")

            if (date != dateOfPreviousEvent) {
                // Reset file content, and prepare the headers for a new CSV file
                Timber.i("csvFile: ${csvFile}")
                stringWriter = StringWriter()
                csvPrinter = CSVPrinter(stringWriter, CSVFormat.DEFAULT.builder().setHeader(
                    "id",
                    "timestamp",
                    "package_name",
                    "mastery_score",
                    "time_spent_ms",
                    "additional_data",
                    "research_experiment",
                    "experiment_group",
                    "letter_sound_letters",
                    "letter_sound_sounds",
                    "letter_sound_id"
                ).get())
            }
            csvPrinter?.printRecord(
                event.id,
                event.timestamp.timeInMillis / 1_000,
                event.packageName,
                event.masteryScore,
                event.timeSpentMs,
                event.additionalData,
                event.researchExperiment?.ordinal,
                event.experimentGroup?.ordinal,
                event.letterSoundLetters,
                event.letterSoundSounds,
                event.letterSoundId
            )
            csvPrinter?.flush()

            // Write the content to the CSV file
            val csvFileContent = stringWriter.toString()
            FileUtils.writeStringToFile(csvFile, csvFileContent, "UTF-8")

            dateOfPreviousEvent = date
        }

        Timber.i("exportLetterSoundAssessmentEvents complete!")
    }

    
    private fun exportWordLearningEvents() {
        Timber.i("exportWordLearningEvents")

        // Read all the events from the database
        val roomDb = RoomDb.getDatabase(applicationContext)
        val wordLearningEventDao = roomDb.wordLearningEventDao()
        val events = wordLearningEventDao.loadAllOrderedByTimestampAsc()
        Timber.i("events.size: ${events.size}")

        // Generate one CSV file per day of events, e.g:
        //   lang-THA/word-learning-events/5b7c682a12ecbe2e_4000021_word-learning-events_2025-06-29.csv
        //   lang-THA/word-learning-events/5b7c682a12ecbe2e_4000021_word-learning-events_2025-06-30.csv
        var stringWriter: StringWriter? = null
        var csvPrinter: CSVPrinter? = null
        var dateOfPreviousEvent: String? = null
        for (event in events) {
            // Get the event's date in ISO format, e.g. "2025-06-29"
            val date: String = eventDateFormat.format(event.timestamp.time)

            // Prepare the CSV file path
            val languageDir = File(applicationContext.filesDir, "lang-${SharedPreferencesHelper.getLanguage(applicationContext)}")
            val eventsDir = File(languageDir, "word-learning-events")
            val csvFile = File(eventsDir, "${event.androidId}_${BuildConfig.VERSION_CODE}_word-learning-events_${date}.csv")

            if (date != dateOfPreviousEvent) {
                // Reset file content, and prepare the headers for a new CSV file
                Timber.i("csvFile: ${csvFile}")
                stringWriter = StringWriter()
                csvPrinter = CSVPrinter(stringWriter, CSVFormat.DEFAULT.builder().setHeader(
                    "id",
                    "timestamp",
                    "package_name",
                    "additional_data",
                    "research_experiment",
                    "experiment_group",
                    "word_text",
                    "word_id"
                ).get())
            }
            csvPrinter?.printRecord(
                event.id,
                event.timestamp.timeInMillis / 1_000,
                event.packageName,
                event.additionalData,
                event.researchExperiment?.ordinal,
                event.experimentGroup?.ordinal,
                event.wordText,
                event.wordId
            )
            csvPrinter?.flush()

            // Write the content to the CSV file
            val csvFileContent = stringWriter.toString()
            FileUtils.writeStringToFile(csvFile, csvFileContent, "UTF-8")

            dateOfPreviousEvent = date
        }

        Timber.i("exportWordLearningEvents complete!")
    }
    
    private fun exportWordAssessmentEvents() {
        Timber.i("exportWordAssessmentEvents")

        // Read all the events from the database
        val roomDb = RoomDb.getDatabase(applicationContext)
        val wordAssessmentEventDao = roomDb.wordAssessmentEventDao()
        val events = wordAssessmentEventDao.loadAllOrderedByTimestampAsc()
        Timber.i("events.size: ${events.size}")

        // Generate one CSV file per day of events, e.g:
        //   lang-THA/word-assessment-events/5b7c682a12ecbe2e_4000021_word-assessment-events_2025-06-29.csv
        //   lang-THA/word-assessment-events/5b7c682a12ecbe2e_4000021_word-assessment-events_2025-06-30.csv
        var stringWriter: StringWriter? = null
        var csvPrinter: CSVPrinter? = null
        var dateOfPreviousEvent: String? = null
        for (event in events) {
            // Get the event's date in ISO format, e.g. "2025-06-29"
            val date: String = eventDateFormat.format(event.timestamp.time)

            // Prepare the CSV file path
            val languageDir = File(applicationContext.filesDir, "lang-${SharedPreferencesHelper.getLanguage(applicationContext)}")
            val eventsDir = File(languageDir, "word-assessment-events")
            val csvFile = File(eventsDir, "${event.androidId}_${BuildConfig.VERSION_CODE}_word-assessment-events_${date}.csv")

            if (date != dateOfPreviousEvent) {
                // Reset file content, and prepare the headers for a new CSV file
                Timber.i("csvFile: ${csvFile}")
                stringWriter = StringWriter()
                csvPrinter = CSVPrinter(stringWriter, CSVFormat.DEFAULT.builder().setHeader(
                    "id",
                    "timestamp",
                    "package_name",
                    "mastery_score",
                    "time_spent_ms",
                    "additional_data",
                    "research_experiment",
                    "experiment_group",
                    "word_text",
                    "word_id"
                ).get())
            }
            csvPrinter?.printRecord(
                event.id,
                event.timestamp.timeInMillis / 1_000,
                event.packageName,
                event.masteryScore,
                event.timeSpentMs,
                event.additionalData,
                event.researchExperiment?.ordinal,
                event.experimentGroup?.ordinal,
                event.wordText,
                event.wordId
            )
            csvPrinter?.flush()

            // Write the content to the CSV file
            val csvFileContent = stringWriter.toString()
            FileUtils.writeStringToFile(csvFile, csvFileContent, "UTF-8")

            dateOfPreviousEvent = date
        }

        Timber.i("exportWordAssessmentEvents complete!")
    }


    private fun exportNumberLearningEvents() {
        Timber.i("exportNumberLearningEvents")

        // Read all the events from the database
        val roomDb = RoomDb.getDatabase(applicationContext)
        val numberLearningEventDao = roomDb.numberLearningEventDao()
        val events = numberLearningEventDao.loadAllOrderedByTimestampAsc()
        Timber.i("events.size: ${events.size}")

        // Generate one CSV file per day of events, e.g:
        //   lang-THA/number-learning-events/5b7c682a12ecbe2e_4000021_number-learning-events_2025-06-29.csv
        //   lang-THA/number-learning-events/5b7c682a12ecbe2e_4000021_number-learning-events_2025-06-30.csv
        var stringWriter: StringWriter? = null
        var csvPrinter: CSVPrinter? = null
        var dateOfPreviousEvent: String? = null
        for (event in events) {
            // Get the event's date in ISO format, e.g. "2025-06-29"
            val date: String = eventDateFormat.format(event.timestamp.time)

            // Prepare the CSV file path
            val languageDir = File(applicationContext.filesDir, "lang-${SharedPreferencesHelper.getLanguage(applicationContext)}")
            val eventsDir = File(languageDir, "number-learning-events")
            val csvFile = File(eventsDir, "${event.androidId}_${BuildConfig.VERSION_CODE}_number-learning-events_${date}.csv")

            if (date != dateOfPreviousEvent) {
                // Reset file content, and prepare the headers for a new CSV file
                Timber.i("csvFile: ${csvFile}")
                stringWriter = StringWriter()
                csvPrinter = CSVPrinter(stringWriter, CSVFormat.DEFAULT.builder().setHeader(
                    "id",
                    "timestamp",
                    "package_name",
                    "additional_data",
                    "research_experiment",
                    "experiment_group",
                    "number_value",
                    "number_symbol",
                    "number_id"
                ).get())
            }
            csvPrinter?.printRecord(
                event.id,
                event.timestamp.timeInMillis / 1_000,
                event.packageName,
                event.additionalData,
                event.researchExperiment?.ordinal,
                event.experimentGroup?.ordinal,
                event.numberValue,
                event.numberSymbol,
                event.numberId
            )
            csvPrinter?.flush()

            // Write the content to the CSV file
            val csvFileContent = stringWriter.toString()
            FileUtils.writeStringToFile(csvFile, csvFileContent, "UTF-8")

            dateOfPreviousEvent = date
        }

        Timber.i("exportNumberLearningEvents complete!")
    }

    private fun exportNumberAssessmentEvents() {
        Timber.i("exportNumberAssessmentEvents")

        // Read all the events from the database
        val roomDb = RoomDb.getDatabase(applicationContext)
        val numberAssessmentEventDao = roomDb.numberAssessmentEventDao()
        val events = numberAssessmentEventDao.loadAllOrderedByTimestampAsc()
        Timber.i("events.size: ${events.size}")

        // Generate one CSV file per day of events, e.g:
        //   lang-THA/number-assessment-events/5b7c682a12ecbe2e_4000021_number-assessment-events_2025-06-29.csv
        //   lang-THA/number-assessment-events/5b7c682a12ecbe2e_4000021_number-assessment-events_2025-06-30.csv
        var stringWriter: StringWriter? = null
        var csvPrinter: CSVPrinter? = null
        var dateOfPreviousEvent: String? = null
        for (event in events) {
            // Get the event's date in ISO format, e.g. "2025-06-29"
            val date: String = eventDateFormat.format(event.timestamp.time)

            // Prepare the CSV file path
            val languageDir = File(applicationContext.filesDir, "lang-${SharedPreferencesHelper.getLanguage(applicationContext)}")
            val eventsDir = File(languageDir, "number-assessment-events")
            val csvFile = File(eventsDir, "${event.androidId}_${BuildConfig.VERSION_CODE}_number-assessment-events_${date}.csv")

            if (date != dateOfPreviousEvent) {
                // Reset file content, and prepare the headers for a new CSV file
                Timber.i("csvFile: ${csvFile}")
                stringWriter = StringWriter()
                csvPrinter = CSVPrinter(stringWriter, CSVFormat.DEFAULT.builder().setHeader(
                    "id",
                    "timestamp",
                    "package_name",
                    "mastery_score",
                    "time_spent_ms",
                    "additional_data",
                    "research_experiment",
                    "experiment_group",
                    "number_value",
                    "number_symbol",
                    "number_id"
                ).get())
            }
            csvPrinter?.printRecord(
                event.id,
                event.timestamp.timeInMillis / 1_000,
                event.packageName,
                event.masteryScore,
                event.timeSpentMs,
                event.additionalData,
                event.researchExperiment?.ordinal,
                event.experimentGroup?.ordinal,
                event.numberValue,
                event.numberSymbol,
                event.numberId
            )
            csvPrinter?.flush()

            // Write the content to the CSV file
            val csvFileContent = stringWriter.toString()
            FileUtils.writeStringToFile(csvFile, csvFileContent, "UTF-8")

            dateOfPreviousEvent = date
        }

        Timber.i("exportNumberAssessmentEvents complete!")
    }


    private fun exportStoryBookLearningEvents() {
        Timber.i("exportStoryBookLearningEvents")

        // Read all the events from the database
        val roomDb = RoomDb.getDatabase(applicationContext)
        val storyBookLearningEventDao = roomDb.storyBookLearningEventDao()
        val events = storyBookLearningEventDao.loadAllOrderedByTimestampAsc()
        Timber.i("events.size: ${events.size}")

        // Generate one CSV file per day of events, e.g:
        //   lang-THA/storybook-learning-events/5b7c682a12ecbe2e_4000021_storybook-learning-events_2025-06-29.csv
        //   lang-THA/storybook-learning-events/5b7c682a12ecbe2e_4000021_storybook-learning-events_2025-06-30.csv
        var stringWriter: StringWriter? = null
        var csvPrinter: CSVPrinter? = null
        var dateOfPreviousEvent: String? = null
        for (event in events) {
            // Get the event's date in ISO format, e.g. "2025-06-29"
            val date: String = eventDateFormat.format(event.timestamp.time)

            // Prepare the CSV file path
            val languageDir = File(applicationContext.filesDir, "lang-${SharedPreferencesHelper.getLanguage(applicationContext)}")
            val eventsDir = File(languageDir, "storybook-learning-events")
            val csvFile = File(eventsDir, "${event.androidId}_${BuildConfig.VERSION_CODE}_storybook-learning-events_${date}.csv")

            if (date != dateOfPreviousEvent) {
                // Reset file content, and prepare the headers for a new CSV file
                Timber.i("csvFile: ${csvFile}")
                stringWriter = StringWriter()
                csvPrinter = CSVPrinter(stringWriter, CSVFormat.DEFAULT.builder().setHeader(
                    "id",
                    "timestamp",
                    "package_name",
                    "additional_data",
                    "research_experiment",
                    "experiment_group",
                    "storybook_title",
                    "storybook_id"
                ).get())
            }
            csvPrinter?.printRecord(
                event.id,
                event.timestamp.timeInMillis / 1_000,
                event.packageName,
                event.additionalData,
                event.researchExperiment?.ordinal,
                event.experimentGroup?.ordinal,
                event.storyBookTitle,
                event.storyBookId
            )
            csvPrinter?.flush()

            // Write the content to the CSV file
            val csvFileContent = stringWriter.toString()
            FileUtils.writeStringToFile(csvFile, csvFileContent, "UTF-8")

            dateOfPreviousEvent = date
        }

        Timber.i("exportStoryBookLearningEvents complete!")
    }


    private fun exportVideoLearningEvents() {
        Timber.i("exportVideoLearningEvents")

        // Read all the events from the database
        val roomDb = RoomDb.getDatabase(applicationContext)
        val videoLearningEventDao = roomDb.videoLearningEventDao()
        val events = videoLearningEventDao.loadAllOrderedByTimestampAsc()
        Timber.i("events.size: ${events.size}")

        // Generate one CSV file per day of events, e.g:
        //   lang-THA/video-learning-events/5b7c682a12ecbe2e_4000021_video-learning-events_2025-06-29.csv
        //   lang-THA/video-learning-events/5b7c682a12ecbe2e_4000021_video-learning-events_2025-06-30.csv
        var stringWriter: StringWriter? = null
        var csvPrinter: CSVPrinter? = null
        var dateOfPreviousEvent: String? = null
        for (event in events) {
            // Get the event's date in ISO format, e.g. "2025-06-29"
            val date: String = eventDateFormat.format(event.timestamp.time)

            // Prepare the CSV file path
            val languageDir = File(applicationContext.filesDir, "lang-${SharedPreferencesHelper.getLanguage(applicationContext)}")
            val eventsDir = File(languageDir, "video-learning-events")
            val csvFile = File(eventsDir, "${event.androidId}_${BuildConfig.VERSION_CODE}_video-learning-events_${date}.csv")

            if (date != dateOfPreviousEvent) {
                // Reset file content, and prepare the headers for a new CSV file
                Timber.i("csvFile: ${csvFile}")
                stringWriter = StringWriter()
                csvPrinter = CSVPrinter(stringWriter, CSVFormat.DEFAULT.builder().setHeader(
                    "id",
                    "timestamp",
                    "package_name",
                    "additional_data",
                    "research_experiment",
                    "experiment_group",
                    "video_title",
                    "video_id"
                ).get())
            }
            csvPrinter?.printRecord(
                event.id,
                event.timestamp.timeInMillis / 1_000,
                event.packageName,
                event.additionalData,
                event.researchExperiment?.ordinal,
                event.experimentGroup?.ordinal,
                event.videoTitle,
                event.videoId
            )
            csvPrinter?.flush()

            // Write the content to the CSV file
            val csvFileContent = stringWriter.toString()
            FileUtils.writeStringToFile(csvFile, csvFileContent, "UTF-8")

            dateOfPreviousEvent = date
        }

        Timber.i("exportVideoLearningEvents complete!")
    }
}
